### DISCUSSION TITLE
Sharing behind reverse proxy with oauth

### DISCUSSION THREAD

**Discussion by robflate** [photoprism/photoprism#705] [Q&A] [answered]
I have Photoprism behind Traefik reverse proxy with oauth from Google via [traefik-forward-auth](https://hub.docker.com/r/thomseddon/traefik-forward-auth). I'd like to enable sharing without requiring oauth or Photoprism login.

To bypass oauth for things that should be public I usually use some combination of the following labels to check for a known path and API key;

```- "traefik.http.routers.app.rule=Host(`app.example.com`) && PathPrefix(`s`,`/api`) && Headers(`X-Api-Key`, `xxxxxxxxxx`)"```
or
```- "traefik.http.routers.app.rule=Host(`app.example.com`) && PathPrefix(`s`,`/api`) && Headers(`Authorization`, `Token xxxxxxxxxx`)"```
or
```- "traefik.http.routers.app.rule=Host(`app.example.com`) && PathPrefix(`s`,`/api`) && Query(`apikey`, `xxxxxxxxxx`)"```

As far as I can tell, there's no master API key in the Photoprism API, instead relying on a request coming from a logged in user?

A less secure alternative would be for me to just bypass any request to `/s` and `/api` but if I do that and go to `https://photoprism.example.com/api` I am forwarded to `https://photoprism.example.com/login` and oauth does not kick in allowing access to the Photoprism login form. This is unusual to me because if I go directly to `https://photoprism.example.com/login` oauth does kick in.

Anyway, long story short. Is anyone using a similar setup that has oauth and managing to successfully use sharing without requiring any auth. Thanks.

**Comment by lastzero** [comment, marked as answer]
Sounds like this specific solution requires code changes... the tokens we use right now are essentially user specific session ids in the HTTP header that may also expire. Note that we don't use cookies to be (client) device agnostic.

**Comment by robflate** [comment]
My post sounds like I'm being prescriptive. I intended it more as a request to see how others were combining security and opening up the sharing side of Photoprism.

I've been messing with some APIs recently, namely;

* [Strapi](https://hub.docker.com/r/strapi/strapi) - Headless CMS to build powerful APIs with no effort
* [Baserow](https://hub.docker.com/r/tiredofit/baserow) - Low/No code data management application
* [SeaTable](https://hub.docker.com/r/seatable/seatable) - Easy like a spreadsheet, powerful like a database
* [Directus](https://hub.docker.com/r/directus/directus) - CMS & API for managing agnostic content 
* [Cockpit](https://hub.docker.com/r/agentejo/cockpit/) - Simple Content Platform to manage any structured content 
* [Ghost](https://hub.docker.com/_/ghost/) - Free and open source blogging platform written in JavaScript

I don't begin to know how this all works other than a very superficial interest but they all take a similar but different approach;

1. API Key attached to user, discoverable in UI and set in Query or X-Api-Key header
2. Temporary JWT or Bearer tokens requested with GET or POST etc
3. Authorization Header

When testing stuff as a single user of a service, option 1 is obviously the simplest but the other options, particularly the temporary tokens must be more secure and scalable to more users?

Anyway, might you know why in my particular case a direct request to `https://photoprism.example.com/login` will correctly invoke oauth and redirect to a Google login page but going to `https://photoprism.example.com/api` will redirect to `https://photoprism.example.com/login` without invoking oauth and thus exposing the Photoprism login page? I have a similar setup on many other containers and this doesn't appear to happen. I understand this is probably entirely outside of Photoprism but just wondered if anything special happens on the redirect from `/api` to `/login` and why the redirect happens at all. Thanks.
